### PR TITLE
Building scenarios

### PR DESCRIPTION
**Pull request by znicholls**
# Pull request

Please confirm that this pull request has done the following:

- [x] Tests added
- [x] Documentation added (where applicable)
- [x] Example added (either to an existing notebook or as a new notebook, where applicable)
- [x] Description in ``CHANGELOG.rst`` added

## Adding to CHANGELOG.rst

Please add a single line in the changelog notes similar to one of the following:

```
- (`#XX <http://link-to-pr.com>`_) Added feature which does something
- (`#XX <http://link-to-pr.com>`_) Fixed bug identified in (`#XX <http://link-to-issue.com>`_)
```

### CODE REVIEW

**Comment by lewisjared** [comment]
rebase this as a lot is already committed to master

**Comment by znicholls** [comment]
Done

File: CHANGELOG.rst [hunk 1]
```diff
@@ -4,6 +4,7 @@ Changelog
 master
 ------
 
+- (`#187 <https://github.com/openclimatedata/pymagicc/pull/187>`_) Added `pymagicc.io.join_timeseries`
```
Comment (rgieseke) [inline CHANGELOG.rst]: Maybe add something like "simplifies joining/merging scenarios to create custom scenarios"?

**Comment by rgieseke** [comment]
The "Building Scenarios" is super useful, but do we have some documentation on the "todo" stuff already somewhere? Might be useful to elaborate or add a  link to docs where this is explained a bit more.

File: pymagicc/io.py [hunk 2]
```diff
@@ -2095,3 +2119,106 @@ def get_generic_rcp_name(inname):
     except KeyError:
         error_msg = "No generic name for input: {}".format(inname)
         raise ValueError(error_msg)
+
+
+# TODO: move out of here in clean up
```
Comment (rgieseke) [inline pymagicc/io.py]: Create an issue, do it or somehow make it clearer how this could be re-factored?

File: pymagicc/io.py [hunk 3]
```diff
@@ -2095,3 +2119,106 @@ def get_generic_rcp_name(inname):
...
+def join_timeseries(base, overwrite, join_linear=None):
```
Comment (rgieseke) [inline pymagicc/io.py]: `join_linear` - should this be a tuple (while the code would still accept a list) or rather, `start`, `stop` - i think we usually have start and end points as separate args (correct me if i'm wrong or if there is a specific reason to have a list)

Comment (znicholls) [reply pymagicc/io.py]: I think having separate start and end is a bit dangerous as if you have a start arg but no end, I'm not sure what should happen... Tuple is good, have altered

**Comment by znicholls** [comment]
> Might be useful to elaborate or add a link to docs where this is explained a bit more.

Nice! Have done (although link won't work until docs are built...)

File: docs/file_conventions.rst [hunk 4]
```diff
@@ -41,6 +41,28 @@ each of these files).
...
+- units: the units of the data
+
+- time: the point in time at which the data occurs
+
+    - note that the internal convention is that state variables are start of year values (i.e. 1st January 1990) whilst fluxes are annual averages/midyear values
+
+- set: a MAGICC flag which defines what MAGICC should do with the data in the timeseries
+
+    - SET: set the given variable in the given region to the value provided
+    - ADD: add the data to the given variable in the given region to any already read in data for that variable in that region (not clear when exactly this is taken into account, question to be answered)
```
Comment (rgieseke) [inline docs/file_conventions.rst]: Very useful to have this all written down -- just not clear to me what the comment in parens means ...?

File: docs/file_conventions.rst [hunk 5]
```diff
@@ -41,6 +41,28 @@ each of these files).
...
+    - SUBTRACT: subtract the data to the given variable in the given region from any already read in data for that variable in that region (not clear when exactly this is taken into account, question to be answered)
```
Comment (rgieseke) [inline docs/file_conventions.rst]: subtract "from"?

**Comment by znicholls** [comment]
@lewisjared hopefully this scenario builder should make your life a bit easier in future

File: pymagicc/io.py [hunk 6]
```diff
@@ -2095,3 +2119,106 @@ def get_generic_rcp_name(inname):
...
+def _join_timeseries_mdata(base, overwrite, join_linear):
+    min_year = min(base["time"].min(), overwrite["time"].min())
+    max_year = max(base["time"].max(), overwrite["time"].max())
+    new_index = range(min_year, max_year + 1)
+
+    result = _reindex_df(base, new_index)
+    ow = _reindex_df(overwrite, new_index)
+
+    if result.stack().index.intersection(ow.stack().index).empty:
+        raise ValueError("No overlapping indices, a simple append will do")
```
Comment (lewisjared) [inline pymagicc/io.py]: Should this raise an exception? The function is meant to join time series so why not just return the appended dataframes in this case?

Comment (znicholls) [reply pymagicc/io.py]: My reasoning here was that if you call join_timeseries, you expect there to be some overwriting done and should be told if that's not the case. Maybe the function needs to be renamed to make that clearer?

Comment (znicholls) [reply pymagicc/io.py]: Or just raise a warning instead?

Comment (znicholls) [reply pymagicc/io.py]: It's to guard against the example in the notebook where you think you're joining scenarios, except something stupid has gone wrong (i.e. the todo column isn't behaving) and so you end up not joining anything and just getting a massive dataframe back which you can't use.

Comment (lewisjared) [reply pymagicc/io.py]: Ahh I see. That raises another question - what is the point of N/A in the TODO column if you just set it to set??? 

Perhaps, there needs to be a higher level function which takes scenario filenames and tries to merge them as it is a bit clunky having to load the scenarios and set the TODO column to set and then call join_timeseries

Comment (znicholls) [reply pymagicc/io.py]: > That raises another question - what is the point of N/A in the TODO column if you just set it to set???

I had this idea that cause they were data files, the user should have to choose what to do with them. On reflection, reading in SET makes more sense.

Will tweak and right high level option now

**Comment by znicholls** [comment]
@lewisjared can you take another look?

**Comment by lewisjared** [comment]
Sure if you push them :D

**Comment by znicholls** [comment]
> Sure if you push them :D

fml

**Comment by lewisjared** [comment]
LGTM once you resolve the failed test

**Pull request merged**